src/linear_algebra/tensor_power.lean

```diff
@@ -46,28 +41,243 @@ variables {R : Type*} {M : Type*} [comm_semiring R] [add_comm_monoid M] [module
+  ∀ {a b : graded_monoid (λ ii, ⨂[R] i : ι ii, M)} (h : a.fst = b.fst),
+    reindex R M (equiv.cast $ congr_arg ι h) a.snd = b.snd → a = b
+| ⟨ai, a⟩ ⟨bi, b⟩ := λ (hi : ai = bi) (h : reindex R M _ a = b),
+begin
+  subst hi,
+  simpa using h,
+end
+
+end pi_tensor_product
+
 namespace tensor_power
-open_locale tensor_product
+open_locale tensor_product direct_sum
 open pi_tensor_product
 
 /-- As a graded monoid, `⨂[R]^i M` has a `1 : ⨂[R]^0 M`. -/
 instance ghas_one : graded_monoid.ghas_one (λ i, ⨂[R]^i M) :=
-{ one := tprod R fin.elim0 }
+{ one := tprod R $ @fin.elim0' M }
 
 local notation `ₜ1` := @graded_monoid.ghas_one.one ℕ (λ i, ⨂[R]^i M) _ _
 
-lemma ghas_one_def : ₜ1 = tprod R fin.elim0 := rfl
+lemma ghas_one_def : ₜ1 = tprod R (@fin.elim0' M) := rfl
 
 /-- A variant of `pi_tensor_prod.tmul_equiv` with the result indexed by `fin (n + m)`. -/
 def mul_equiv {n m : ℕ} : (⨂[R]^n M) ⊗[R] (⨂[R]^m M) ≃ₗ[R] ⨂[R]^(n + m) M :=
 (tmul_equiv R M).trans (reindex R M fin_sum_fin_equiv)
 
 /-- As a graded monoid, `⨂[R]^i M` has a `(*) : ⨂[R]^i M → ⨂[R]^j M → ⨂[R]^(i + j) M`. -/
 instance ghas_mul : graded_monoid.ghas_mul (λ i, ⨂[R]^i M) :=
-{ mul := λ i j a b, mul_equiv (a ⊗ₜ b) }
+{ mul := λ i j a b, (tensor_product.mk R _ _).compr₂ ↑(mul_equiv : _ ≃ₗ[R] ⨂[R]^(i + j) M) a b}
 
 local infix ` ₜ* `:70 := @graded_monoid.ghas_mul.mul ℕ (λ i, ⨂[R]^i M) _ _ _ _
 
 lemma ghas_mul_def {i j} (a : ⨂[R]^i M) (b : ⨂[R]^j M) : a ₜ* b = mul_equiv (a ⊗ₜ b) := rfl
 
+lemma ghas_mul_eq_coe_linear_map {i j} (a : ⨂[R]^i M) (b : ⨂[R]^j M) :
+  a ₜ* b =
+    ((tensor_product.mk R _ _).compr₂ ↑(mul_equiv : _ ≃ₗ[R] ⨂[R]^(i + j) M)
+      : ⨂[R]^i M →ₗ[R] ⨂[R]^j M →ₗ[R] ⨂[R]^(i + j) M) a b := rfl
+
+variables (R M)
+
+/-- Cast between "equal" tensor powers. -/
+def cast {i j} (h : i = j) : ⨂[R]^i M ≃ₗ[R] (⨂[R]^j M) :=
+reindex R M (fin.cast h).to_equiv
+
+lemma cast_tprod {i j} (h : i = j) (a : fin i → M) :
+  cast R M h (tprod R a) = tprod R (a ∘ fin.cast h.symm) :=
+reindex_tprod _ _
+
+@[simp] lemma cast_refl {i} (h : i = i) : (cast R M h) = linear_equiv.refl _ _ :=
+begin
+  refine eq.trans _ (reindex_refl),
+  rw cast,
+  refine congr_arg (reindex R M) _,
+  rw fin.cast_refl,
+  refl,
+end
+
+@[simp] lemma cast_symm {i j} (h : i = j) : (cast R M h).symm = cast R M h.symm := reindex_symm _
+
+@[simp] lemma cast_trans {i j k} (h : i = j) (h' : j = k) :
+  (cast R M h).trans (cast R M h') = cast R M (h.trans h') := reindex_trans _ _
+
+variables {R M}
+
+@[simp] lemma cast_cast {i j k} (h : i = j) (h' : j = k) (a : ⨂[R]^i M) :
+  cast R M h' (cast R M h a) = cast R M (h.trans h') a := reindex_reindex _ _ _
+
+@[ext]
+lemma graded_monoid_eq_of_cast {a b : graded_monoid (λ n, ⨂[R] i : fin n, M)}
+  (h : a.fst = b.fst) (h2 : cast R M h a.snd = b.snd) : a = b :=
+begin
+  refine graded_monoid_eq_of_reindex_cast h _,
+  rw cast at h2,
+  rw [←fin.cast_to_equiv, ← h2],
+end
+
+lemma cast_eq_cast {i j} (h : i = j) : ⇑(cast R M h) = _root_.cast (congr_arg _ h) :=
```

Comment:
```suggestion
@[norm_cast] lemma coe_cast_eq_cast {i j} (h : i = j) : ⇑(cast R M h) = _root_.cast (congr_arg _ h) :=
```

Reply:
This is intended to match `fin.cast_eq_cast`, which is not norm_cast. I've added a comment making the link.